Only derive JsonError for errors that can return Json

DanGould opened the pull request:
- 1st commit de-duplicates
- 2nd commit: Only some errors are actually recoverable and must return JSON. Fix https://github.com/payjoin/rust-payjoin/issues/522.

Re: #403

DanGould commented:
My refactor separates replyable from unreplyable errors. The biggest hurdle I've run into is when the payload gets parsed as a string before `validate_payload`. That prevents the `PayloadError::Utf8` from being shared between v1 and v2 payload parsing even though the receiver has the sender's reply key at this point and could in fact reply to the sender. Going to think for a moment about whether or not a different format could be used to serialize the `psbt || \n || query` payload that'd get around this problem. Perhaps even serializing the query as a field of a strict byte length in `message_a` instead of in the string payload.

spacebear21 commented on payjoin/src/receive/v2/mod.rs:
```diff
@@ -158,14 +157,8 @@ impl Receiver {
-            url::form_urlencoded::parse(query.as_bytes()),
-            SUPPORTED_VERSIONS,
-        )
-        .map_err(InternalPayloadError::SenderParams)?;
```
on line 182 before this change, there is a `log::debug!` that is now duplicated in `validate_payload`.

spacebear21 commented on payjoin/src/receive/mod.rs:
```diff
@@ -60,3 +63,22 @@ impl InputPair {
+    base64: String,
+    query: &str,
+    supported_versions: &'static [usize],
+) -> Result<(Psbt, Params), PayloadError> {
```
This could return a `Result<v1::UncheckedProposal, PayloadError>`.

DanGould replied:
Not sure about that. V2 needs to disable_output_substitution for v1 backwards-compatibility. Exclusive v1 does not.

spacebear21 replied:
Good point. Perhaps `validate_payload` should take responsibility for doing this backwards compatibility check if v2 is enabled? (I might be stretching the meaning of `validate` here.)

DanGould replied:
To do that you'd need a feature flag in the now named parse_payload fn which kind of defeats the purpose of the shared fn

spacebear21 commented on payjoin-cli/src/app/v1.rs:
```diff
@@ -16,7 +16,7 @@ use hyper_util::rt::TokioIo;
 use payjoin::bitcoin::{self, FeeRate};
 use payjoin::receive::v1::{PayjoinProposal, UncheckedProposal};
-use payjoin::receive::Error;
+use payjoin::receive::ReplyableError;
```
For brevity I prefer importing `ReplyableError::{V1, Implementation}` and using the variants directly (like it's done in `app/v2.rs`).

spacebear21 commented on payjoin-cli/src/app/v2.rs:
```diff
@@ -337,25 +334,27 @@ async fn handle_request_error(
     ohttp_relay: &payjoin::Url,
 ) -> anyhow::Error {
-    let (err_req, err_ctx) = match receiver.extract_err_req(&e, ohttp_relay) {
+    if let Some((err_req, err_ctx)) = match receiver.extract_err_req(&e, ohttp_relay) {
```
I don't understand this `if let Some`, doesn't the `match` return a Result?

DanGould replied:
no, the initial block unwraps the Ok and convernts the err to anyhow!, then the Some value gets used in the second block

spacebear21 replied:
ohh I missed that the signature of `extract_err_req` was changed to return an Option 🤦

spacebear21 reviewed (commented):
What do you think about naming the variants `RecoverableError`/`UnrecoverableError` instead of `Un/Replyable`? My brain keeps auto-correcting that to "replayable".

DanGould commented:
I've changed the design again.

This time, Error is dividend into non_exhaustive Recoverable and V2 variants. I note that I do still prefer `Replyable` since some of the V2 variants are recoverable in the sense that they just require different configuration, they don't cause the program to crash. Jsonable or ReturnToSender names could also work.

I removed the `Unrecoverable` variant since it only covered v2::SessionError types, and the `receive::Error` type is almost certainly going to be extended with more multiparty variants.

I'm not sure if it makes sense for the v1 module to return `RecoverableError`, as it does now, since all the errors are indeed recoverable, or if it should be wrapped in the single receive::Error layer since that would let it match the other v2 module.

@spacebear21 please let me know what you think. I do indeed believe this is much closer to a stable reality based on actual hierarchy than it was before, especially since the fundamental difference between errors that may be returned to the sender is separated.

spacebear21 reviewed (commented):
I'm ok with `ReturnToSender` or maybe `Returnable`.

> I'm not sure if it makes sense for the v1 module to return RecoverableError

I'm not sure exactly where you mean - in `extract_proposal_from_v1`? I'm fine with returning the lowest common denominator (RecoverableError) in v1 functions and letting callers map to `Error` if needed.

On another note, I wonder if we should box a `ImplementationError = Box<dyn Error + Send + Sync>` type so that the receiver function signatures can be more semantically correct:

e.g.

```rust
        pub fn identify_receiver_outputs(
        self,
        is_receiver_output: impl Fn(&Script) -> Result<bool, ImplementationError>,
    ) -> Result<WantsOutputs, RecoverableError> {
```

DanGould commented:
Regarding names, I took a happy medium keeping ReplyableError but using the ReplyToSender variant.

Re:
> we should box a ImplementationError = Box<dyn Error + Send + Sync> type so that the receiver function signatures can be more semantically correct:

This is an incredible idea! I was playing around with returning Box<dyn Error> from these functions yesterday but didn't quite discover that an alias is how to make this semantically correct. Added it as a third commit and it looks great, way easier to implement the receive state machine properly by using `ImplementationError::from`.

Now I'm wondering how to handle the straggling receive errors: SelectionError, OutputSubstitutionError, InputContributionError. If unrecoverable, they're probably all ReplyableError variants with an `unavailable` or `not enough money` error code. I think I know how to address it in a follow up.

spacebear21 reviewed (approved):
ACK 1358c50d797fc7f1b10de70d0fe8fe35ed724329

> Now I'm wondering how to handle the straggling receive errors: SelectionError, OutputSubstitutionError, InputContributionError. If unrecoverable, they're probably all ReplyableError variants with an unavailable or not enough money error code. I think I know how to address it in a follow up.

I think these are all technically recoverable from the receiver side, e.g. "try again with a different input selection". If the receiver can't handle them then they become unavailable in most cases. I'm not sure we should use the `not-enough-money` error code for `InternalInputContributionError::ValueTooLow` because BIP78 says it specifically applies to "could not bump the _fee_ of the payjoin proposal. "
